pkg/backup/compaction_test.go

```diff
@@ -556,6 +548,187 @@ func TestBlockConcurrentScheduledCompactions(t *testing.T) {
+				}},
+			},
+			2: {
+				ExternalIODir: "/east0",
+				Locality: roachpb.Locality{Tiers: []roachpb.Tier{
+					{Key: "tier", Value: "0"},
+					{Key: "region", Value: "east"},
+				}},
+			},
+			3: {
+				ExternalIODir: "/east1",
+				Locality: roachpb.Locality{Tiers: []roachpb.Tier{
+					{Key: "tier", Value: "1"},
+					{Key: "region", Value: "east"},
+				}},
+			},
+		},
+	}
+	const numAccounts = 1000
+	tc, db, _, cleanup := backupRestoreTestSetupWithParams(t, 4, numAccounts, InitManualReplication, args)
+	defer cleanup()
+
+	west1Node, east0Node := sqlutils.MakeSQLRunner(tc.Conns[1]), sqlutils.MakeSQLRunner(tc.Conns[2])
+
+	targets := "DATABASE data"
+
+	// bring all node dirs up to an identical starting point
+	// with one full backup and 3 uncompacted incrementals.
+	// returns start and end timestamps
+	bringUp := func(subCollection string) (hlc.Timestamp, hlc.Timestamp) {
+		start := getTime()
+		// initial full backups
+		for i := 1; i <= 4; i++ {
+			db.Exec(t, fullBackupQuery(targets,
+				[]string{fmt.Sprintf("nodelocal://%d/%s", i, subCollection)},
+				start, noOpts))
+		}
+		db.Exec(t, "UPDATE data.bank SET balance = 200")
+		for i := 1; i <= 4; i++ {
+			db.Exec(t, incBackupQuery(targets,
+				[]string{fmt.Sprintf("nodelocal://%d/%s", i, subCollection)},
+				noAOST, noOpts))
+		}
+		db.Exec(t, "UPDATE data.bank SET balance = 201")
+		for i := 1; i <= 4; i++ {
+			db.Exec(t, incBackupQuery(targets,
+				[]string{fmt.Sprintf("nodelocal://%d/%s", i, subCollection)},
+				noAOST, noOpts))
+		}
+		db.Exec(t, "UPDATE data.bank SET balance = 202")
+		end := getTime()
+		for i := 1; i <= 4; i++ {
+			db.Exec(t, incBackupQuery(targets,
+				[]string{fmt.Sprintf("nodelocal://%d/%s", i, subCollection)},
+				end, noOpts))
+		}
+		return start, end
+	}
+	numInitialBackups := 4
+
+	countBackups := func(collection string) int {
+		var count int
+		db.QueryRow(
+			t, fmt.Sprintf(
+				"SELECT count(DISTINCT (start_time, end_time)) FROM [SHOW BACKUP FROM LATEST IN '%s']",
+				collection,
+			),
+		).Scan(&count)
+		return count
+	}
+
+	t.Run("pin-tier", func(t *testing.T) {
+		ensureLeaseholder(t, db)
+		start, end := bringUp("pin-tier")
+		processorCollection := []string{"nodelocal://0/pin-tier"}
+
+		// run and wait for locality filtered compaction job
+		jobutils.WaitForJobToSucceed(
+			t, db,
+			triggerCompaction(
+				t, east0Node,
+				fullBackupQuery(targets, processorCollection, start, "WITH EXECUTION LOCALITY = 'tier=0'"),
+				getLatestFullDir(t, db, processorCollection[0]),
+				start, end,
+			),
+		)
+
+		numWest0 := countBackups("nodelocal://1/pin-tier")
+		numWest1 := countBackups("nodelocal://2/pin-tier")
+		numEast0 := countBackups("nodelocal://3/pin-tier")
+		numEast1 := countBackups("nodelocal://4/pin-tier")
+
+		// check that at least one node matching locality filter has all the backups, including compacted
+		require.True(t, numEast0 == numInitialBackups+1 || numWest0 == numInitialBackups+1)
```

Comment:
This does make me wonder if there's an easy way to determine if new files were written to local storage for the non-coordinator node, but probably not. Just some musings though, I don't think we need to do that here.